synapse/metrics/shared_usage_metrics.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2022 The Matrix.org Foundation C.I.C
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import TYPE_CHECKING
+
+from synapse.metrics.background_process_metrics import run_as_background_process
+
+if TYPE_CHECKING:
+    from synapse.server import HomeServer
+
+from prometheus_client import Gauge
+
+# Gauge to expose daily active users metrics
+current_dau_gauge = Gauge(
+    "synapse_admin_daily_active_users",
+    "Current daily active users count",
+)
+
+
+class SharedUsageMetrics:
+    """Usage metrics shared between the phone home stats and the prometheus exporter."""
+
+    def __init__(self, hs: "HomeServer") -> None:
+        self._store = hs.get_datastores().main
+        self._clock = hs.get_clock()
+
+        self.daily_active_users = -1
```

Comment:
I'm not a fan of having each field have a sentinel value; it gives the impression that we may end up reporting `-1` (although I see that you update upon initialisation). IMO we shouldn't report anything rather than report negative values. I can see this being potentially troublesome for the phone-home stats when we sum DAUs.

I think I would be tempted to restructure this a bit:

- rename to `SharedUsageMetricsManager` or equivalent (not sure what the conventional term is in Synapse ... is it `Controller`?)
- make `SharedUsageMetrics` a plain dataclass/attrs object, with `daily_active_users` as an int field
- put a `latest(): Optional[SharedUsageMetrics]` method (or field maybe?) on `SharedUsageMetricsManager`


(N.B. I was going to suggest that the metrics be lazily computed, either on-phone-home or on-prometheus-scrape, but I realise we don't want to block the prometheus scrape endpoint on Twisted's reactor in case everything is going awry)

This sort of design prevents having individual metrics having sentinel values like `-1`, which may escape.